blogposts/inside-rust/01-announcement.md

```diff
@@ -0,0 +1,166 @@
+are also part of the C++ language, which introduced them to the world of
+unmanaged systems programming.
+
+When an exception is thrown, the runtime _unwinds_ the stack, essentially
+traversing it backwards and calling clean-up or error-recovery code such as
+destructors or `catch` blocks.
+
+Compilers may implement their own unwinding mechanisms, but in native code such
+as Rust binaries, the mechanism is more commonly provided by the platform ABI.
+
+It is well known that Rust does not have exceptions as such. But Rust _does_
+support unwinding! There are two scenarios that will cause unwinding to occur:
+
+* By default, Rust's `panic!()` unwinds the stack.
+* Using FFI, Rust can call functions in other languages (such as C++) that can
+  unwind the stack.
+  * There are some special cases where C libraries can actually cause
+    unwinding.  For instance, on Microsoft platforms, `longjmp` is implemented
+    as "forced unwinding" (see below)
+
+Currently, when foreign (non-Rust) code invokes Rust code, the behavior of a
+`panic!()` unwind that "escapes" from Rust is explicitly undefined. Similarly,
+when Rust calls a foreign function that unwinds, the behavior once the unwind
+operation encounters Rust frames is undefined. The primary reason for this is
+to ensure that Rust implementations may use their own unwinding mechanism,
+which may not be compatible with the platform-provided "native" unwinding
+mechanism. Currently, however, `rustc` uses the native mechanism, and there are
+no plans to change this.
+
+### Forced unwinding
+
+Platform ABIs can define a special kind of unwinding called "forced unwinding."
+This type of unwinding operation cannot be stopped prematurely using language
+mechanisms such as Rust's `catch_unwind` or C++'s `catch`. (These mechanisms
+can intercept forced unwinds, but the runtime will resume unwinding regardless
+of the content of the intercepting code, unless the program is aborted).
+
+There are two common types of forced unwinding:
+
+* On Windows platforms, `longjmp` is implemented as a forced unwind.
+* On UNIX-like platforms, `pthread_exit` results in a forced unwind.
+
+## Requirements for any cross-language unwinding specification
+
+* Unwinding between Rust functions (and in particular unwinding of Rust panics)
+  may not necessarily use the system unwinding mechanism
+  * In practice, we do use the system mechanism today, but we would like to
+    reserve the freedom to change this
+* If you enable `-Cpanic=abort`, we are able to optimize the size of binaries
+  to remove most code related to unwinding.
+  * In extreme cases, it should be possible to remove **all** traces of
+    unwinding, when you know that it will not occur.
+  * In practice, most "C" functions are never expected to unwind (because they
+    are written in C, for example, and not in C++).
+    * However, because unwinding is now part of most system ABIs, even C
```

Comment:
Does anyone know whether this would be a "forced unwind" in practice? i.e., is this the same as `pthread_exit`? (maybe @gnzlbg or @Amanieu knows?)

Reply:
Yes, syscall cancellation due to `pthread_cancel` is basically a forced unwind. It is exactly the same mechanism as the one used by `pthread_exit`.